spark/src/main/scala/com/gu/tableversions/spark/SparkHiveMetastore.scala

```diff
@@ -0,0 +1,183 @@
+    for {
+      tableLocation <- findTableLocation(table).map(new URI(_))
+      partitionLocation = partitionVersion.partition.resolvePath(tableLocation)
+      versionedPartitionLocation = VersionPaths.pathFor(partitionLocation, partitionVersion.version)
+    } yield versionedPartitionLocation
+
+  private def updateTableLocation(table: TableName, tableLocation: URI, version: VersionNumber): F[Unit] = {
+    val versionedPath = VersionPaths.pathFor(tableLocation, version)
+    val updateQuery = s"ALTER TABLE ${table.fullyQualifiedName} SET LOCATION '$versionedPath'"
+    performUpdate(s"Updating table location of table ${table.fullyQualifiedName}", updateQuery)
+  }
+
+  private def performUpdate(description: String, query: String): F[Unit] =
+    F.delay {
+      logger.info(s"$description using query: $query")
+      spark.sql(query)
+    }.void
+
+  private def findTableLocation(table: TableName): F[String] = {
+    val query = s"DESCRIBE FORMATTED ${table.fullyQualifiedName}"
+
+    F.delay(spark.sql(query).where('col_name === "Location").collect().toList.map(_.getString(1)).headOption)
+      .map(_.getOrElse(throw new Exception(s"No location information returned for table ${table.fullyQualifiedName}")))
+  }
+
+  private def listPartitions(table: TableName): F[List[String]] =
+    F.delay { spark.sql(s"show partitions ${table.fullyQualifiedName}").collect().toList.map(_.getString(0)) }
+
+  private def partitionLocation(table: TableName, partitionExpr: String): F[String] =
+    F.delay {
+        spark
+          .sql(s"DESCRIBE FORMATTED ${table.fullyQualifiedName} PARTITION $partitionExpr")
+          .where('col_name === "Location")
+          .collect()
+          .map(_.getAs[String]("data_type"))
+          .headOption
+      }
+      .map(_.getOrElse(throw new Exception(
+        s"No location information returned for partition $partitionExpr on table ${table.fullyQualifiedName}")))
+
+  private def isPartitioned(table: TableName): F[Boolean] = {
+    // We have to interpret the strange format returned by DESCRIBE queries, which, if the table is partitioned,
+    // contains rows like:
+    //
+    // |# Partition Information|         |               |
+    // |# col_name             |data_type|comment        |
+    // |date                   |date     |null           |
+
+    F.delay {
+      spark
+        .sql(s"DESCRIBE ${table.fullyQualifiedName}")
+        .collect()
+        .flatMap(row => Option(row.getAs[String]("col_name")))
+        .contains("# Partition Information")
+    }
+  }
+
+}
+
+object SparkHiveMetastore {
+
+  private[spark] def toPartitionExpr(partitionPath: String): String =
+    toHivePartitionExpr(parsePartition(partitionPath))
+
+  private[spark] def toHivePartitionExpr(partition: Partition): String =
+    partition.columnValues
+      .map(columnValue => s"${columnValue.column.name}='${columnValue.value}'")
+      .mkString("(", ",", ")")
+
+  private val ColumnValueRegex = "([a-z_]+)=(.+)".r
```

Comment:
Minor, but I like the `x` flag to allow comments in regexes longer than more than a few characters. For example, this could be written as: 

```suggestion
  private val ColumnValueRegex = """
      |(?x)
      |([a-z_]+)  # column name
      |=
      |(.+)       # column value
    """.stripMargin
```

Reply:
TIL!

Reply:
Got it to work with some minor adjustments (the `(?x)` has to be before any newlines, and there's the `.r`):

```scala
  private val ColumnValueRegex = """(?x)
                                   |([a-z_]+)  # column name
                                   |=
                                   |(.+)       # column value
                                 """.stripMargin.r
```